### PR TITLE
Start working on tests

### PR DESCRIPTION
**Pull request by ThePhilgrim**
@Akuli I pushed what I have so you have the chance to debug if you get the time.

Todo list (feel free to edit):
- [x] Edit `.github/workflows/github_actions.yml` to also run pytest
- [x] Edit `.github/workflows/github_actions.yml` to run `black` on all python files, `mypy` on all files except tests
- [x] Create a second test. There will be a lot of copy/pasta, use a pytest fixture to clean it up.

### CODE REVIEW

File: tests/test_mantatail.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    client_socket = socket.socket()
+    client_socket.connect(("localhost", 6667))
+    client_socket.sendall(b"NICK foo\r\n")
+
+    # Receiving everything the server is going to send helps prevent errors.
+    # Otherwise it might not be fully started yet when the client quits.
+    received = b''
+    while not received.endswith(b'End of /MOTD command\r\n'):
+        received += client_socket.recv(100)
```
Comment (ThePhilgrim) [inline tests/test_mantatail.py:26]: From Python docs:
"Note For best match with hardware and network realities, the value of bufsize should be a relatively small power of 2, for example, 4096."

https://docs.python.org/3/library/socket.html#socket.socket.recv

```suggestion
        received += client_socket.recv(4096)
```

File: mantatail.py [hunk 2]
```diff
@@ -214,6 +218,16 @@ def recv_loop(self, user: User, command_handler: IrcCommandHandler) -> None:
                     call_handler_function(message)
 
 
+def get_motd_content_from_json() -> Union[Dict[str, List[str]], str]:
+    try:
+        with open("./resources/motd.json", "r") as file:
+            motd_content: Dict[str, List[str]] = json.load(file)
+            return motd_content
+    except FileNotFoundError:
+        return ":MOTD File is missing\r\n"
```
Comment (Akuli) [inline mantatail.py:227]: I think it's better if the server just doesn't start when the MOTD file is missing. You will quicker notice that something is wrong. This simplifies some things.

File: tests/test_mantatail.py [hunk 3]
```diff
@@ -21,11 +33,11 @@ def run_server(server):
 
     # Receiving everything the server is going to send helps prevent errors.
     # Otherwise it might not be fully started yet when the client quits.
-    received = b''
-    while not received.endswith(b'End of /MOTD command\r\n'):
+    received = b""
+    while not received.endswith(b"- End test MOTD\r\n"):
```
Comment (Akuli) [inline tests/test_mantatail.py:37]: This should wait for `376 End of /MOTD command`, which comes after `372 End test MOTD`.

File: mantatail.py [hunk 4]
```diff
@@ -50,27 +50,30 @@ def handle_motd(self) -> None:
...
-        motd = motd_content["motd"]
 
         self.user.socket.sendall(start_msg)
 
-        for motd_line in motd:
-            motd_msg = bytes(
-                f"{self.send_to_client_prefix} {motd_num} {self.user.nick} :{motd_line.format(user_nick=self.user.nick)}{self.send_to_client_suffix}",
-                encoding=self.encoding,
+        if self.server.motd_content:
+            print("if")
```
Comment (Akuli) [inline mantatail.py:59]:
```suggestion
```

File: mantatail.py [hunk 5]
```diff
@@ -50,27 +50,30 @@ def handle_motd(self) -> None:
...
+            motd = self.server.motd_content["motd"]
+            for motd_line in motd:
+                motd_msg = bytes(
+                    f"{self.send_to_client_prefix} {motd_num} {self.user.nick} :{motd_line.format(user_nick=self.user.nick)}{self.send_to_client_suffix}",
+                    encoding=self.encoding,
+                )
+                self.user.socket.sendall(motd_msg)
+        # If motd.json could not be found
+        else:
+            print("else")
```
Comment (Akuli) [inline mantatail.py:69]:
```suggestion
```

File: tests/test_mantatail.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                nonlocal last_exception
+                last_exception = e
+
+    monkeypatch.setattr(threading, "Thread", ThreadWrapper)
+    yield
+    if last_exception:
+        raise last_exception
+
+
+@pytest.fixture(autouse=True, scope="function")
```
Comment (Akuli) [inline tests/test_mantatail.py:41]: `scope="function"` is the default, no need to change:

```suggestion
@pytest.fixture(autouse=True)
```

File: tests/test_mantatail.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    # .shutdown() raises an OSError on mac, removing it makes the test suite freeze on linux.
+    try:
+        server.listener_socket.shutdown(socket.SHUT_RDWR)
+    except OSError:
+        pass
+
+    server.listener_socket.close()
+
+
+@pytest.fixture(scope="function")
```
Comment (Akuli) [inline tests/test_mantatail.py:63]:
```suggestion
@pytest.fixture
```

File: tests/test_mantatail.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    received = b""
+    while not received.endswith(b"\r\n:mantatail 376 alice :End of /MOTD command\r\n"):
+        received += alice_socket.recv(4096)
+
+    yield alice_socket
+    alice_socket.sendall(b"QUIT\r\n")
+    alice_socket.close()
+
+
+@pytest.fixture(scope="function")
```
Comment (Akuli) [inline tests/test_mantatail.py:80]:
```suggestion
@pytest.fixture
```

**Pull request merged**